daemons/ipa-kdb/ipa_kdb_kdcpolicy.c

```diff
@@ -44,13 +45,90 @@ ipa_kdcpolicy_check_as(krb5_context context, krb5_kdcpolicy_moddata moddata,
+
     *status = NULL;
     *lifetime_out = 0;
     *renew_lifetime_out = 0;
 
     krb5_klog_syslog(LOG_INFO, "IPA kdcpolicy: checking AS-REQ.");
 
-    return 0;
+    ied = (struct ipadb_e_data *)client->e_data;
+    if ((ied == NULL) || (ied->magic != IPA_E_DATA_MAGIC)) {
+        /* e-data is not availble, getting user auth from LDAP */
+        krb5_klog_syslog(LOG_INFO, "IPA kdcpolicy: client e_data not availble. Try fetching...");
+        kerr = ipadb_get_principal(context, request->client, KRB5_KDB_FLAG_ALIAS_OK, &client);
+        if (kerr != 0) {
+            krb5_klog_syslog(LOG_ERR, "IPA kdcpolicy: ipadb_find_principal failed.");
+            goto done;
+        }
+
+        ied = (struct ipadb_e_data *)client->e_data;
+        if ((ied == NULL) && (ied->magic != IPA_E_DATA_MAGIC)) {
+            krb5_klog_syslog(LOG_ERR, "IPA kdcpolicy: client e_data fetching failed.");
+            kerr = EINVAL;
+            goto done;
+        }
+    }
+
+    ua = ied->user_auth;
+    
+    /* If no mechanisms are set, allow every auth method */
+    if (ua == IPADB_USER_AUTH_NONE) {
+        goto done;
+    }
+
+    /* For each auth indicator, see if it is allowed for that user */
+    for (int i = 0; auth_indicators[i] != NULL; i++) {
+        const char *auth_indicator = auth_indicators[i];
+
+        if (strcmp(*auth_indicators, "otp") == 0) {
+            valid_auth_indicators++;
+            if (!(ua & IPADB_USER_AUTH_OTP)) {
+                kerr = KRB5KDC_ERR_POLICY;
+                *status = "OTP pre-authentication not allowed for this user.";
+                goto done;
+            }
+
+        } else if (strcmp(*auth_indicators, "radius") == 0) {
+            valid_auth_indicators++;
+            if (!(ua & IPADB_USER_AUTH_RADIUS)) {
+                kerr = KRB5KDC_ERR_POLICY;
+                *status = "OTP pre-authentication not allowed for this user.";
+                goto done;
+            }
+
+        } else if (strcmp(*auth_indicators, "pkinit") == 0) {
+            valid_auth_indicators++;
+            if (!(ua & IPADB_USER_AUTH_PKINIT)) {
+                kerr = KRB5KDC_ERR_POLICY;
+                *status = "PKINIT pre-authentication not allowed for this user.";
+                goto done;
+            }
+
+        } else if (strcmp(*auth_indicators, "hardened") == 0) {
+            valid_auth_indicators++;
+            /* Allow hardened even if only password pre-auth is allowed */
+            if (!(ua & (IPADB_USER_AUTH_HARDENED | IPADB_USER_AUTH_PASSWORD))) {
+                kerr = KRB5KDC_ERR_POLICY;
+                *status = "Password pre-authentication not not allowed for this user.";
+                goto done;
+            }
+        }
+    }
+
+    if (!valid_auth_indicators) {
+        if (!(ua & IPADB_USER_AUTH_PASSWORD)) {
+            kerr = KRB5KDC_ERR_POLICY;
+            *status = "Non-hardened password authentication not allowed for this user.";
+        }
+    }
+
+done:
```

Comment:
If all that's happening in this label is a `return`, don't bother with the label - there's no cleanup, so it's easier to follow if you just inline the return.